scli-runner/src/main/scala/com.olegych.scastie.sclirunner/ScliRunner.scala

```diff
@@ -58,45 +70,85 @@ class ScliRunner {
 
-        val charOffsetInstrumentation = userDirectives.map(_.length() + 1).sum
+    val code = allDirectives.mkString("\n") + "\n" + inputs.code
+    writeFile(scalaMain, code)
 
-        val code = allDirectives.mkString("\n") + "\n" + inputs.code
-        writeFile(scalaMain, code)
+    var instrumentationMem: Option[List[Instrumentation]] = None
+    var outputBuffer: List[String] = List()
 
-        var instrumentationMem: Option[List[Instrumentation]] = None
+    val build = bspClient.build(snippetId.base64UUID)
+    build.map { bspBuild =>
+      log.info(s"Running process for snippet $snippetId")
 
-        val build = bspClient.build(task.snippetId.base64UUID, line => {
-          extract[List[Instrumentation]](line) match { // extract instrumentation
-            case None => onOutput(line)
+      def forwardPrint(str: String) = {
+        onOutput(str)
+        outputBuffer = str :: outputBuffer
+      }
+        
+      val runProcess = bspBuild.process.run(
+        ProcessLogger({ line: String => {
+          // extract instrumentation
+          extract[List[Instrumentation]](line) match {
+            case None => forwardPrint(line)
             case Some(value) => {
               instrumentationMem = Some(value.map(inst => inst.copy(
                 position = inst.position.copy(inst.position.start + charOffsetInstrumentation, inst.position.end + charOffsetInstrumentation)
               )))
             }
           }
-        })
-
-        build map { r =>
-          r.copy(instrumentation = instrumentationMem)
-        } recover {
-          case x: BspClient.CompilationError => throw CompilationError(x.toProblemList.map(pb =>
-            pb.copy(line = pb.line.map(_ + totalOffset + 1))))
-          case other => throw other
-        } 
+
+        }})
+      )
+      javaProcesses.put(snippetId, runProcess)
+
+      // Wait for process to exit gracefully (or being killed by timeout…)
+      val f = Future { runProcess.exitValue() }
+      Try(Await.result(f, Duration(30, TimeUnit.SECONDS))) match {
+        case Success(value) => true
+        case Failure(_: TimeoutException) => {
+          forwardPrint("Timeout exceeded.")
+          false
+        }
+        case Failure(e) => {
+          forwardPrint(s"Unknown exception $e")
+          false
+        }
       }
-    }
+    } map { didFail =>
+      if (didFail) {
+        javaProcesses.get(snippetId).map(_.destroy())
+      }
+      javaProcesses.remove(snippetId)
+
+      ScliRun(output = outputBuffer, instrumentationMem)
+    } recover {
+      case x: BspClient.CompilationError => throw CompilationError(x.toProblemList.map(pb =>
+        pb.copy(line = pb.line.map(_ + totalOffset + 1))))
+      case other => throw other
+    } 
   }
 
   def end: Unit = {
     bspClient.end
+    javaProcesses.values.foreach(_.destroy())
     process.map(_.destroy())
   }
 
+  // Java processes
+  private val javaProcesses = HashMap[SnippetId, Process]() // mutable HashMap
```

Comment:
HashMap is not concurrent safe. And I think we are going to support multiple runtimes at the same time. Try to use `TrieMap` instead.

Reply:
Done!